lxd/patches.go

```diff
@@ -1845,4 +1848,152 @@ func patchDefaultStoragePermissions(_ string, d *Daemon) error {
 	return nil
 }
 
+// patchUpdateVolatileAttachedVolumesFormat updates "volatile.attached_volumes" from old format (map of volume UUID -> snapshot UUID) to new format (map of device_name -> snapshot_UUID).
+func patchUpdateVolatileAttachedVolumesFormat(_ string, d *Daemon) error {
+	s := d.State()
+
+	// Only run on a single cluster member to avoid concurrent updates.
+	isSelectedMember, err := selectedPatchClusterMember(s)
+	if err != nil {
+		return err
+	}
+
+	if !isSelectedMember {
+		return nil
+	}
+
+	err = s.DB.Cluster.Transaction(s.ShutdownCtx, func(ctx context.Context, tx *db.ClusterTx) error {
+		type snapshotData struct {
+			snapshotID                   int
+			instanceID                   int
+			projectName                  string
+			volatileAttachedVolumesValue string
+		}
+
+		var snapshots []snapshotData
+
+		// Query to get all instance snapshots with "volatile.attached_volumes".
+		q := `
+SELECT 
+	instances_snapshots.id,
+	instances_snapshots.instance_id,
+	projects.name,
+	instances_snapshots_config.value
+FROM instances_snapshots_config
+JOIN instances_snapshots ON instances_snapshots.id = instances_snapshots_config.instance_snapshot_id
+JOIN instances ON instances.id = instances_snapshots.instance_id
+JOIN projects ON projects.id = instances.project_id
+WHERE instances_snapshots_config.key = "volatile.attached_volumes"
+`
+		err := query.Scan(ctx, tx.Tx(), q, func(scan func(dest ...any) error) error {
+			var snap snapshotData
+
+			err := scan(&snap.snapshotID, &snap.instanceID, &snap.projectName, &snap.volatileAttachedVolumesValue)
+			if err != nil {
+				return err
+			}
+
+			snapshots = append(snapshots, snap)
+
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+
+		for _, snap := range snapshots {
+			var attachedVolumesData map[string]string
+			err = json.Unmarshal([]byte(snap.volatileAttachedVolumesValue), &attachedVolumesData)
+			if err != nil || len(attachedVolumesData) == 0 {
+				if err != nil {
+					logger.Warn(`Failed parsing "volatile.attached_volumes"`, logger.Ctx{"err": err})
+				}
+
+				continue
+			}
+
+			// Get instance snapshot devices.
+			devices, err := dbCluster.GetInstanceSnapshotDevices(ctx, tx.Tx(), snap.snapshotID)
+			if err != nil {
+				return fmt.Errorf("Failed getting instance snapshot devices: %w", err)
+			}
+
+			// Skip if already in new format (any key matches a device name).
+			isNewFormat := false
+			for key := range attachedVolumesData {
+				_, ok := devices[key]
+				if ok {
+					isNewFormat = true
+					break
+				}
+			}
+
+			if isNewFormat {
+				continue
+			}
+
+			// Convert from old format (volume UUID -> snapshot UUID) to new format (device name -> snapshot UUID).
+			newVolatileAttachedVolumes := make(map[string]string)
+			for name, dev := range devices {
+				if !filters.IsCustomVolumeDisk(dev.Config) {
+					continue
+				}
+
+				poolID, err := tx.GetStoragePoolID(ctx, dev.Config["pool"])
```

Comment:
Can we push this down into a single query by taking the list of volume uuids keys from `attachedVolumesData` and then running a query that gets the volumes referenced by those UUIDs, along with their pool name and volume name.

Then you can convert that list to a pool name & vol name -> vol uuid map, and use that to find the device(s) that they match to.

BTW I thought of some things to check:

1. If a project doesn't have the `features.storage.volumes` enabled, do the UUIDs point to volumes in the default project, rather than the instance project?
2. Can a single volume be used by multiple instance devices (yes), so this needs to be taken into account when converting the devices (if not already).

Reply:
Great, thanks!